File: frontend/src/types/vector.ts
/** Shapes returned by GET /api/v1/measurements/uploads/{id}/vector. */

export interface VectorBlock {
  block_index: number;
  start_sample: number;
  time_s: number;
  amplitude: number;
  /** Raw atan2 result; absolute only in an arbitrary sense — see phase_reference. */
  phase_deg: number;
  /** Phase against block 0. This is the meaningful quantity. */
  relative_phase_deg: number;
  is_reference: boolean;
}

export interface VectorDrift {
  phase_slope_deg_per_s: number | null;
  implied_frequency_offset_hz: number | null;
  amplitude_cv: number | null;
  linear_fit_r2: number | null;
  likely_bin_mismatch: boolean;
}

export interface FrequencyCandidate {
  label: string;
  frequency_hz: number;
  source: string;
  amplitude: number | null;
}

export interface VibrationVectorResponse {
  upload_id: string;
  sensor_id: string;
  channel: number;
  captured_at: string;
  original_filename: string | null;

  sampling_rate_hz: number;
  sample_count: number;
  window: string;
  block_size: number;
  block_step: number;
  block_count: number;
  overlap: number;
  block_duration_s: number;

  target_hz_requested: number;
  bin_hz: number;
  bin_index: number;
  frequency_resolution_hz: number;
  target_source: string;
  target_label: string;

  amplitude_unit: string;
  data_type: string;
  amplitude_min: number;
  amplitude_max: number;

  /** FFT-derived estimate. Never a measured tachometer reading. */
  estimated_shaft_hz: number | null;

  phase_reference: string;
  angle_convention: string;
  keyphasor_available: boolean;

  sensor_label: string | null;
  orientation: string | null;
  mounting_location: string | null;

  blocks: VectorBlock[];
  drift: VectorDrift;
  candidates: FrequencyCandidate[];
}

export interface VibrationVectorQuery {
  uploadId: string;
  channel: number;
  targetHz?: number;
  blockSize?: number;
  overlap?: number;
}

/** Block sizes offered in the UI. Larger blocks resolve low frequencies more precisely. */
export const VECTOR_BLOCK_SIZES = [512, 1024, 1600, 2048, 4096, 8192] as const;

File: backend/app/schemas/vector.py
"""Response models for the polar / vibration vector plot."""
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class VectorBlockOut(BaseModel):
    """One FFT analysis block of a single capture."""
    block_index: int
    start_sample: int
    time_s: float
    amplitude: float
    #: Raw atan2 result. Absolute only in the arbitrary sense — see phase_reference.
    phase_deg: float
    #: Phase relative to block 0. This is the meaningful quantity.
    relative_phase_deg: float
    is_reference: bool


class VectorDriftOut(BaseModel):
    """Informational bin-mismatch heuristic — never a fault verdict."""
    phase_slope_deg_per_s: Optional[float] = None
    implied_frequency_offset_hz: Optional[float] = None
    amplitude_cv: Optional[float] = None
    linear_fit_r2: Optional[float] = None
    likely_bin_mismatch: bool = False


class FrequencyCandidateOut(BaseModel):
    """A frequency the user can pick as the order of interest."""
    label: str
    frequency_hz: float
    #: "estimated_shaft" (1x/2x/3x) or "peak" (from the stored spectrum).
    source: str
    amplitude: Optional[float] = None


class VibrationVectorOut(BaseModel):
    upload_id: UUID
    sensor_id: UUID
    channel: int
    captured_at: datetime
    original_filename: Optional[str] = None

    sampling_rate_hz: float
    sample_count: int
    window: str
    block_size: int
    block_step: int
    block_count: int
    overlap: float
    block_duration_s: float

    target_hz_requested: float
    bin_hz: float
    bin_index: int
    frequency_resolution_hz: float
    target_source: str
    target_label: str

    amplitude_unit: str
    data_type: str
    amplitude_min: float
    amplitude_max: float

    #: FFT-derived estimate, never a measured tachometer reading. May be null.
    estimated_shaft_hz: Optional[float] = None

    phase_reference: str
    angle_convention: str
    keyphasor_available: bool = False

    sensor_label: Optional[str] = None
    orientation: Optional[str] = None
    mounting_location: Optional[str] = None

    blocks: List[VectorBlockOut]
    drift: VectorDriftOut
    candidates: List[FrequencyCandidateOut] = Field(default_factory=list)
